### PR TITLE
feat(common): add extra overload to async pipe

### LINKED ISSUE #36404: add Promise<T> | Observable<T> async pipe overload
# 🚀 feature request

### Relevant Package
`@angular/common`

### Description
Currently the async pipe can take a Promise or Observable, but only as seperate obverloads, so with `strictTemplates` you get an error if you pass in a `Promise<T> | Observable<T>`

### Describe the solution you'd like
Add an overload which looks like `transform<T>(obj: Promise<T> | Observable<T> | null | undefined): T | null;`

### Describe alternatives you've considered
converting the promise into a observable, but this isn't ergonomic

Happy to submit a patch

### PR DESCRIPTION
**Pull request by Toxicable**
## PR Checklist
- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
- [x] Feature

## What is the current behavior?
Issue Number: #36404

## What is the new behavior?
```
The async pipes supports inputs being
Promise or Observable, this adds an overload
which is a union of both Promise and Observable
```

## Does this PR introduce a breaking change?

- [x] No

### CODE REVIEW

**Comment by Toxicable** [comment]
I don't understand why there's linting issues with code I han't touched.
I can run the linter on them anyway but it would be confusing for the reviewer to see all this random linting changes.

**Comment by petebacondarwin** [comment]
@Toxicable - I suspect it was that you needed to rebase because the formatter was picking up differences in old commits? I see you have just pushed again so hopefully that will sort it.

**Comment by JoostK** [comment]
The unrelated formatting issues are due to #36203, where clang-format was updated that affects the expected file format. I wonder if it wouldn't be better to just bite the bullet of reformatting all sources across the repo to avoid spurious changes for PRs, but it would probably mean lots of conflicts for existing PRs.

**Comment by Toxicable** [comment]
@JoostK Ahh thanks for pointing to that.
What i've done in our repository at work when making changes like this, is to as you mentioned, do the global change, it's a little bit of pain up front and at least it's clear for devs on what they need to do.

Whatever the decision is there I don't think we'll make the changes in this PR, however to pass CI looks like I have to run the linter on the changed files.

So I applogise to the reviewer for the extra noise in the diff

**Comment by Toxicable** [comment]
The test failures look like flakes

File: goldens/public-api/common/common.d.ts [hunk 1]
```diff
@@ -7,6 +7,7 @@ export declare class AsyncPipe implements OnDestroy, PipeTransform {
     transform<T>(obj: undefined): undefined;
     transform<T>(obj: Observable<T> | null | undefined): T | null;
```
Comment (Airblader) [inline goldens/public-api/common/common.d.ts:8]: The single overloads are no longer needed, see they?

**Review by jelbourn — APPROVED** [review]
LGTM

Reviewed-for: public-api

(I realize this PR hasn't been touched in a while, but the bot just added me this week)

**Review by petebacondarwin — APPROVED** [review]
Reviewed-for: public-api

I am not sure why the test failed but it could be valid. I couldn't seem to re-run it.
@Toxicable perhaps you could trigger a rebuild by rebasing?

File: packages/common/src/pipes/async_pipe.ts [hunk 2]
```diff
@@ -88,6 +99,7 @@ export class AsyncPipe implements OnDestroy, PipeTransform {
   transform<T>(obj: undefined): undefined;
   transform<T>(obj: Observable<T>|null|undefined): T|null;
   transform<T>(obj: Promise<T>|null|undefined): T|null;
+  transform<T>(obj: Observable<T>|Promise<T>|null|undefined): T|null;
```
Comment (mhevery) [inline packages/common/src/pipes/async_pipe.ts:102]: Remove previous two overloads as they provide no value.

File: packages/common/src/pipes/async_pipe.ts [hunk 3]
```diff
@@ -19,17 +19,28 @@ interface SubscriptionStrategy {
 
 class ObservableStrategy implements SubscriptionStrategy {
   createSubscription(async: Observable<any>, updateLatestValue: any): SubscriptionLike {
-    return async.subscribe({next: updateLatestValue, error: (e: any) => { throw e; }});
+    return async.subscribe({
```
Comment (mhevery) [inline packages/common/src/pipes/async_pipe.ts:22]: revert these changes as they are formatting only?

**Review by mhevery — CHANGES_REQUESTED** [review]

**Comment by atscott** [comment]
This would be superseded by #37447.

**Review by atscott — APPROVED** [review]
Reviewed-for: public-api

As mentioned before, #37447 would be a more complete fix but is blocked until v11 since it's breaking. This seems good as an intermediate, non-breaking change.

**Comment by AndrewKushnir** [comment]
Closing this PR as the type improvement was implemented in #37447.
